File: python/app/ingest.py
import json
import os
import sys
from dataclasses import dataclass
from datetime import datetime

@dataclass(frozen=True)
class Play:
    ts: datetime
    ms_played: int
    track: str
    artist: str
    album: str
    track_uri: str

def load_json_file(file_path):
    """
    loads JSON file into python list
    """
    if not os.path.exists(file_path):
        print(f"Error: the file '{file_path}' does not exist")
        return None
    try:
        with open(file_path, 'r', encoding='utf-8') as file:
            data = json.load(file)
            return data

    except json.JSONDecodeError:
        print(f"Error: '{file_path}' is not valid JSON")
        return None

    except Exception as e:  # pylint: disable=broad-exception-caught
        # Intentional catch-all: any unexpected I/O error here should be
        # reported to the user, not crash the CLI.
        print(f"An unexpected error occurred: {e}")
        return None

def parse_timestamp(raw):
    if raw is None:
        return None
    return datetime.fromisoformat(raw)

def parse_entry(entry):
    return Play(
        ts= parse_timestamp(entry.get("ts")),
        ms_played= entry.get("ms_played"),
        track= entry.get("master_metadata_track_name"),
        artist= entry.get("master_metadata_album_artist_name"),
        album= entry.get("master_metadata_album_album_name"),
        track_uri= entry.get("spotify_track_uri")
    )

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python ingest.py <path-to-json>")
        sys.exit(1)

    records = load_json_file(sys.argv[1])
    if records is None:
        sys.exit(1)

    for r in records:
        print(parse_entry(r))

File: go/internal/ingest/ingest.go
// Package ingest parses a Spotify Extended Streaming History export into
// Play records.
//
// Ports python/app/ingest.py:
//   - Play struct       (ts, msPlayed, track, artist, album, trackURI)
//   - LoadJSONFile(path string) ([]map[string]any, error)
//   - ParseTimestamp(raw string) (time.Time, error)
//   - ParseEntry(entry map[string]any) (Play, error)
//
// Test fixture at testdata/spotify_sample.json (copied from
// python/tests/fixtures/spotify_sample.json).
package ingest

import (
	"encoding/json"
	"os"
	"time"
)

type Play struct {
	TS       time.Time
	MSPlayed int
	Track    string
	Artist   string
	Album    string
	TrackURI string
}

func LoadJSONFile(filePath string) ([]map[string]any, error) {
	file, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}

	defer func() { _ = file.Close() }()

	var records []map[string]any
	if err := json.NewDecoder(file).Decode(&records); err != nil {
		return nil, err
	}
	return records, nil
}

func ParseEntry(entry map[string]any) Play {
	tsRaw, _ := entry["ts"].(string)
	ts, _ := time.Parse(time.RFC3339, tsRaw)
	msPlayed, _ := entry["ms_played"].(float64)
	track, _ := entry["master_metadata_track_name"].(string)
	artist, _ := entry["master_metadata_album_artist_name"].(string)
	album, _ := entry["master_metadata_album_album_name"].(string)
	trackURI, _ := entry["spotify_track_uri"].(string)

	return Play{
		TS:       ts,
		MSPlayed: int(msPlayed),
		Track:    track,
		Artist:   artist,
		Album:    album,
		TrackURI: trackURI,
	}
}
